src/conn_poll.c

```diff
@@ -31,33 +40,196 @@ typedef struct registry_impl {
 
 typedef enum conn_state { CONN_STATE_NEW = 0, CONN_STATE_READY, CONN_STATE_FINISHED } conn_state_t;
 
+typedef enum poll_entry_type {
+	POLL_ENTRY_INTERRUPT = 0,
+	POLL_ENTRY_UDP,
+	POLL_ENTRY_TCP_LISTENER,
+	POLL_ENTRY_TCP_CONN,
+} poll_entry_type_t;
+
+typedef struct tcp_connection {
+	socket_t sock;
+	addr_record_t remote;
+	bool closed;
+	bool connecting;
+	uint16_t frame_len;
+	size_t header_bytes;
+	size_t payload_len;
+	uint8_t *send_buffer;
+	size_t send_size;
+	size_t send_offset;
+	size_t send_capacity;
+	uint8_t recv_buffer[65535];
+} tcp_connection_t;
+
 typedef struct conn_impl {
 	conn_registry_t *registry;
 	conn_state_t state;
-	socket_t udp_sock;
-	socket_t tcp_sock;
-	tcp_ice_write_context_t tcp_ice_write_context;
-	tcp_ice_read_context_t tcp_ice_read_context;
-	addr_record_t tcp_dst;
-	tcp_state_t tcp_state;
-	uint16_t ice_tcp_len;
+	socket_t sock;
 	mutex_t send_mutex;
 	int send_ds;
 	timestamp_t next_timestamp;
+	bool tcp_enabled;
+	socket_t tcp_listener;
+	tcp_connection_t *tcp_conns;
+	size_t tcp_conns_count;
+	size_t tcp_conns_capacity;
+	addr_record_t *tcp_addrs;
+	size_t tcp_addrs_count;
 } conn_impl_t;
 
+typedef struct tcp_listener_config {
+	const char *bind_address;
+	uint16_t port_begin;
+	uint16_t port_end;
+} tcp_listener_config_t;
+
+static uint16_t conn_poll_get_next_port(uint16_t begin, uint16_t end) {
```

Comment:
This should be in `tcp.c`. Also, it is a duplicate of `get_next_port_in_range`, so you should reuse existing code instead.